pywt/_multilevel.py

```diff
@@ -397,3 +413,174 @@ def iswt2(coeffs, wavelet):
+        [ 1.,  1.,  1.,  1.]],
+
+       [[ 1.,  1.,  1.,  1.],
+        [ 1.,  1.,  1.,  1.],
+        [ 1.,  1.,  1.,  1.],
+        [ 1.,  1.,  1.,  1.]]])
+
+    """
+    data = np.asarray(data)
+
+    if len(data.shape) < 1:
+        raise ValueError("Expected at least 1D input data.")
+
+    if not isinstance(wavelet, Wavelet):
+        wavelet = Wavelet(wavelet)
+
+    level = _check_level(min(data.shape), wavelet.dec_len, level)
+    coeffs_list = []
+
+    a = data
+    for i in range(level):
+        coeffs = dwtn(a, wavelet, mode)
+        a = coeffs.pop('a' * data.ndim)
+        coeffs_list.append(coeffs)
+
+    coeffs_list.append(a)
+    coeffs_list.reverse()
+
+    if len(coeffs_list) < level + 1:
+        raise ValueError("incomplete coefficient list encountered")
+
+    return coeffs_list
+
+
+def _match_coeff_dims(a_coeff, d_coeff_dict):
+    # For each axis, compare the approximation coeff shape to one of the
+    # stored detail coeffs and truncate the last element along the axis
+    # if necessary.
+    if a_coeff is None:
+        return None
+    if not d_coeff_dict:
+        return a_coeff
+    d_coeff = d_coeff_dict[next(iter(d_coeff_dict))]
+    size_diffs = np.subtract(a_coeff.shape, d_coeff.shape)
+    if np.any((size_diffs < 0) | (size_diffs > 1)):
+        raise ValueError("incompatible coefficient array sizes")
+    return a_coeff[[slice(s) for s in d_coeff.shape]]
+
+
+def waverecn(coeffs, wavelet, mode='symmetric'):
+    """
+    Multilevel nD Inverse Discrete Wavelet Transform.
+
+    coeffs : array_like
+        Coefficients list [cAn, {details_level_n}, ... {details_level_1}]
+    wavelet : Wavelet object or name string
+        Wavelet to use
+    mode : str, optional
+        Signal extension mode, see MODES (default: 'sym')
+
+    Returns
+    -------
+    nD array of reconstructed data.
+
+    Examples
+    --------
+    >>> from pywt import multilevel
+    >>> coeffs = multilevel.wavedecn(np.ones((4, 4, 4)), 'db1')
+    >>> # Levels:
+    >>> len(coeffs)-1
+    2
+    >>> multilevel.waverecn(coeffs, 'db1')
+    array([[[ 1.,  1.,  1.,  1.],
+        [ 1.,  1.,  1.,  1.],
+        [ 1.,  1.,  1.,  1.],
+        [ 1.,  1.,  1.,  1.]],
+
+       [[ 1.,  1.,  1.,  1.],
+        [ 1.,  1.,  1.,  1.],
+        [ 1.,  1.,  1.,  1.],
+        [ 1.,  1.,  1.,  1.]],
+
+       [[ 1.,  1.,  1.,  1.],
+        [ 1.,  1.,  1.,  1.],
+        [ 1.,  1.,  1.,  1.],
+        [ 1.,  1.,  1.,  1.]],
+
+       [[ 1.,  1.,  1.,  1.],
+        [ 1.,  1.,  1.,  1.],
+        [ 1.,  1.,  1.,  1.],
+        [ 1.,  1.,  1.,  1.]]])
+    """
+
+    if not isinstance(coeffs, (list, tuple)):
+        raise ValueError("Expected sequence of coefficient arrays.")
+
+    if len(coeffs) < 1:
+        raise ValueError(
+            "Coefficient list too short (minimum 1 array required).")
+
+    a, ds = coeffs[0], coeffs[1:]
+
+    # Raise error for invalid key combinations
+    ds = list(map(_fix_coeffs, ds))
```

Comment:
Also need need to test that the length of all coeffs are equal to `a.ndim`, otherwise the output has the wrong shape.